src/parse.rs

```diff
@@ -621,6 +650,41 @@ fn parse_decimal(s: &str, ix: usize) -> Option<(usize, usize)> {
     usize::from_str(&s[ix..end]).ok().map(|val| (end, val))
 }
 
+// finds the an ID between < and > and returns (id, skip) where skip
+// is how far ahead we should jump ix (how much of the string we used)
+// IDs have to be made up of word characters (\w)
+fn parse_id(s: &str, ix: usize) -> Option<(&str, usize)> {
+    // find the start of the ID
+    let mut start = 0;
+    for (i, c) in s[ix..].char_indices() {
+        if c == '<' {
+            start = ix + i + 1;
+            break;
+        }
+    }
+    if start == 0 {
+        return None;
+    }
+    
+    // look for the end of the ID while making sure the chars are \w
+    let mut end = 0;
```

Comment:
I think we could write the below in a more concise way:

```
    if let Some(len) = s[start..].find(|c: char| !c.is_alphanumeric() && c != '_') {
        let end = start+len;
        if len > 0 && s[end..].starts_with('>') {
            Some((&s[start..end], end-ix+1))
        } else {
            None
        }
    } else {
        None
    }
```

Reply:
Very cool. I didn't know about closures as Patterns for find.